apm-agent-plugins/apm-es-restclient-plugin/apm-es-restclient-plugin-5_6/src/main/java/co/elastic/apm/agent/es/restclient/v5_6/ElasticsearchClientAsyncInstrumentation.java

```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return named("org.elasticsearch.client.RestClient").
+            and(not(
+                declaresMethod(named("performRequestAsync")
+                    .and(takesArguments(2)
+                        .and(takesArgument(0, named("org.elasticsearch.client.Request")))))));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("performRequestAsync")
+            .and(takesArguments(7)
+                .and(takesArgument(4, named("org.elasticsearch.client.HttpAsyncResponseConsumerFactory")))
+                .and(takesArgument(5, named("org.elasticsearch.client.ResponseListener"))));
+    }
+
+
+    private static class ElasticsearchRestClientAsyncAdvice {
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void onBeforeExecute(@Advice.Argument(0) String method,
+                                            @Advice.Argument(1) String endpoint,
+                                            @Advice.Argument(3) @Nullable HttpEntity entity,
+                                            @Advice.Argument(value = 5, readOnly = false) ResponseListener responseListener,
+                                            @Advice.Local("span") Span span,
+                                            @Advice.Local("wrapped") boolean wrapped,
+                                            @Advice.Local("helper") ElasticsearchRestClientInstrumentationHelper<HttpEntity, Response, ResponseListener> helper) {
+
+            helper = esClientInstrHelperManager.getForClassLoaderOfClass(Response.class);
+            if (helper != null) {
+                span = helper.createClientSpan(method, endpoint, entity);
+                if (span != null) {
+                    responseListener = helper.<ResponseListener>wrapResponseListener(responseListener, span);
+                    wrapped = true;
+                }
+            }
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+        private static void onAfterExecute(@Advice.Argument(5) ResponseListener responseListener,
+                                           @Advice.Local("span") @Nullable Span span,
+                                           @Advice.Local("wrapped") boolean wrapped,
+                                           @Advice.Local("helper") @Nullable ElasticsearchRestClientInstrumentationHelper<HttpEntity, Response, ResponseListener> helper, @Advice.Thrown @Nullable Throwable t) {
+            if (span != null) {
+                // Deactivate in this thread. Span will be ended and reported by the listener
+                span.deactivate();
+
+                if (!wrapped) {
+                    // Listener is not wrapped- we need to end the span so to avoid leak and report error if occurred during method invocation
```

Comment:
I might be wrong but if there was an exception in the method invocation, wrapped wouldn't wrapped still be `true`? So this is basically only in case `helper.wrapResponseListener()` threw an exception. If that's the case, couldn't we just do an instance-of check and remove the wrapped flag altogether?
```suggestion
                    // Listener is not wrapped- we need to end the span so to avoid leak and report error if occurred in the enter advice
```

Reply:
You are not wrong, this is a state for us to know whether we wrapped the listener or not. If we did not wrap it, we need to end the span properly no matter what and while doing so it would be informative to report an Exception if such occurred **during the method invocation**.

> If that's the case, couldn't we just do an instance-of check and remove the wrapped flag altogether

We couldn't because the `ResponseListenerWrapper` class the advice will see is the one visible to the application, meaning the one loaded (or most likely - failed to be loaded) from BootstrapCL, whereas the one we actually used is loaded from a child CL of the app CL.

We could use `listener.getClass().getName().equals(...)`, but what's the benefit over using a boolean?